Lib/test/test_functools.py

```diff
@@ -403,6 +403,33 @@ def test_attributes_unwritable(self):
         else:
             self.fail('partial object allowed __dict__ to be deleted')
 
+    def test_manually_adding_non_string_keyword(self):
+        p = self.partial(capture)
+        # Adding a non-string/unicode keyword to partial kwargs
+        p.keywords[1] = 10
+        self.assertIn('1=10', repr(p))
+        try:
+            p()
+        except TypeError:
+            pass
+        else:
+            self.fail('partial object allowed passing non-string keywords')
+
+        # Deleting the key-value pair during key formatting shouldn't segfault.
+        class MutatesYourDict(object):
+            def __init__(self, dct):
+                self.dct = dct
+            def __str__(self):
+                del self.dct[self]
```

Comment:
Replacing was kind of weird - I think because it's undefined behaviour I tried to test in this case and replacing could create an object that "fills" the memory location where the dangling value had been. I had some trouble to make that case fail in a predictable way without the `INCREF`s. It sometimes threw an AttributeError instead of segfaulting.